src/java.base/share/classes/javax/crypto/KEM.java

```diff
@@ -0,0 +1,631 @@
+ *    var kp = kpg.generateKeyPair();
+ *
+ *    var kem = KEM.getInstance("DHKEM");
+ *    var sender = kem.newEncapsulator(kp.getPublic());
+ *    var encapsulated = sender.encapsulate();
+ *    var k1 = encapsulated.key();
+ *
+ *    var receiver = kem.newDecapsulator(kp.getPrivate());
+ *    var k2 = receiver.decapsulate(encapsulated.encapsulation());
+ *
+ *    assert Arrays.equals(k1.getEncoded(), k2.getEncoded());
+ * }
+ */
+public final class KEM {
+
+    /**
+     * Type for the return value of the {@link Encapsulator#encapsulate} method.
+     * <p>
+     * Note: the key encapsulation message can be also referred to as ciphertext.
+     *
+     * @param key the shared secret as a key, must not be {@code null}
+     * @param encapsulation the key encapsulation message to be sent to the
+     *                      receiver, must not be {@code null}
+     * @param params additional parameters to be sent to the receiver,
+     *               can be {@code null}. The parameters can be used to
+     *               initialize an {@code AlgorithmParameters} so that the
+     *               receiver may recover the {@code AlgorithmParameterSpec}
+     *               used by the encapsulator.
+     *
+     * @see KEM#newEncapsulator(PublicKey, AlgorithmParameterSpec, SecureRandom)
+     */
+    public record Encapsulated(SecretKey key, byte[] encapsulation, byte[] params) {
+        /**
+         * Constructs an {@code Encapsulated} object.
+         *
+         * @param key the shared secret as a key, must not be {@code null}.
+         * @param encapsulation the key encapsulation message, must not be {@code null}.
+         * @param params additional parameters, can be {@code null}.
+         */
+        public Encapsulated {
+            Objects.requireNonNull(key);
+            Objects.requireNonNull(encapsulation);
+        }
+    }
+
+    /**
+     * A decapsulator, generated by {@link #newDecapsulator}.
+     */
+    public static final class Decapsulator {
+        private final KEMSpi.DecapsulatorSpi d;
+        private final Provider p;
+
+        private Decapsulator(KEMSpi.DecapsulatorSpi d, Provider p) {
+            assert d != null;
+            assert p != null;
+            this.d = d;
+            this.p = p;
+        }
+
+        /**
+         * Returns the provider.
+         *
+         * @return thr provider
+         */
+        public Provider provider() {
+            return p;
+        }
+
+        /**
+         * The key decapsulation function.
+         * <p>
+         * This method is equivalent to
+         * {@code decapsulate(encapsulation, 0, secretSize(), "Generic")}. This
+         * combination of arguments should be supported by every implementation.
+         *
+         * @param encapsulation the key encapsulation message from the sender
+         * @return the shared secret as a {@code SecretKey} with
+         *          algorithm being "Generic", not {@code null}
```

Comment:
See my comment below, I don't think you need to specify that this does not return null, it should be implied.